### PR TITLE
Migration functions: replace_event_type, delete_event_type & delete_events

### PR DESCRIPTION
**Pull request by pelegdvir**
Create some sql functions to help alembic and ansible data migrations:
* `array_sort_unique` - returns sorted unique array (used for updating state)
* `replace_event_type` - replaces event type
* `delete_event_type` - deletes events type (updates parent_id)
* `delete_events` - deletes any events (passed as a temp table - more details about why temp table and not array later)
* `delete_events_with_no_parent_id_constraint` - same as `delete_events` - but used for the models with `mode.__unique_parent__=false`)

Why do we need it?
Writing migrations for microcosm-eventsource is not easy and can easily cause bugs - especially when involving adding or removing events.

Why sql function and not sql alchemy code / python script?
We want to support alembic and ansible data migrations.
I've saw (and wrote) some ansible data migrations that are sql only - and I want to support them.

What next?
If we want to continue with this style, we will add `insert_event` function and some `model_rollup` function.

Why `delete_events` input is a temp table and not an array?
While `delete_events` input is really an array, we're going to use it only for migrations - so we need to think about the migration use case:
1. 
The TEMP TABLE syntax:
```
CREATE TEMP TABLE events_to_remove AS (
    SELECT id FROM task_event WHERE event_type='SCHEDULED'
);
SELECT delete_events('task_event', 'events_to_remove', 'task_id');
```
is much more clear then the syntax that we would use with array:
```
SELECT delete_events(
    'task_event',
    array_agg(task_event.id),
    'task_id'
)
FROM task_event WHERE event_type='SCHEDULED';
```
There is no question about the beginning and end of the transaction - it's clear.

2.
Performance - in order to write `delete_events` - we have to use some joins. So we will have to create temp table anyway.
I'm also not sure uncertain what is the performance hit of passing huge arrays between functions.

### CODE REVIEW

File: microcosm_eventsource/ddl/delete_event_type.create.ddl [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * (Example: for company_event table, the model_id_row_name should be company_id)
+ *
+ * Details:
+ * - Updates parent_id of relevant child events
+ * - Updates the state column (Ensures that the state is sorted and has unique events)
+ * - Does not promise valid events or transitions.
+ * - Cannot be used if model.__unique_parent__ is set to False (missing table_names_parent_id_key constraint),
+ */
+CREATE OR REPLACE FUNCTION delete_event_type(
+    table_name regclass,
```
Comment (pelegdvir) [inline microcosm_eventsource/ddl/delete_event_type.create.ddl]: There is no easy way to pass tables to functions in psql.
So we're passing strings and using `EXECUTE format(...)`
For migrations - it's safe enough.

File: microcosm_eventsource/ddl/delete_event_type.create.ddl [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    old_event_type character varying(255),
+    model_id_row_name character varying(255)) RETURNS void AS
+$func$
+BEGIN
+    EXECUTE format(
+        '
+            CREATE TEMP TABLE events_to_remove_%%1$s_%%2$s AS (
+                SELECT id
+                FROM %%1$s
+                WHERE event_type = ''%%2$s''
```
Comment (pelegdvir) [inline microcosm_eventsource/ddl/delete_event_type.create.ddl]: about `''%%2$s''`:
* `''` is psql escaping for `'`
* `%%` python escaping for `%` (sqlalchemy/alembic creates this function)
* `%2$s`: `format` function style for passing args (similar to c)

In python, the syntax was: `f"WHERE event_type = '{old_event_type}'"`

**Comment by jessemyers** [comment]
Postgres function names tend to be `noun_verb`, e.g. `date_foo` and `array_bar`. What do you think about using `event_` for the naming?

File: microcosm_eventsource/tests/ddl/test_migrations.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+        assert_that(results, contains(
+            has_properties(
+                event_type=TaskEventType.STARTED,
+                state=[TaskEventType.STARTED],
+                id=self.started_event.id,
+                parent_id=self.assigened_event.id,
+            ),
+            has_properties(
+                event_type=TaskEventType.ASSIGNED,
+                state=[TaskEventType.ASSIGNED, TaskEventType.CREATED, TaskEventType.SCHEDULED],
```
[the remark this answers is not in the document]
Comment (pelegdvir) [reply microcosm_eventsource/tests/ddl/test_migrations.py]: No. `delete_events` just deletes the passed events.
`delete_event_type` Will update the state.
Most of the time we will use `delete_event_type`.
I will add some documentation.

**Comment by pelegdvir** [comment]
I agree that we need clear examples. Other then real use, I can add examples to the function descriptions and split the tests into two files - one for edge cases and another for the simple cases - that people actually read.

**Comment by pelegdvir** [comment]
There is no official standard.
Also nothin in the psql example (`void` function called `clean_emp` https://www.postgresql.org/docs/8.3/static/xfunc-sql.html)
The best thing that I've found is http://www.nyx.net/~bwunder/dbChangeControl/standard.htm - `proc_` prefix  

So - i've used: `proc_event_type_delete` and `proc_events_delete`
